nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/ListenSlack.java

```diff
@@ -0,0 +1,199 @@
+    MATCHED_MESSAGES_RELATIONSHIP, UNMATCHED_MESSAGES_RELATIONSHIP)));
+
+  @Override
+  public Set<Relationship> getRelationships() {
+    return RELATIONSHIPS;
+  }
+
+  @Override
+  public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+    return DESCRIPTORS;
+  }
+
+  @OnScheduled
+  public void onScheduled(final ProcessContext context) {
+    String messageTypes = context.getProperty(MESSAGE_TYPES).getValue();
+    messageTypes = messageTypes == null ? "" : messageTypes;
+    matchAny = messageTypes.isEmpty();
+    matchingTypes = Arrays.asList(messageTypes.split(","));
+  }
+
+  @OnUnscheduled
+  @OnShutdown
+  public void onUnscheduled() {
+    if (isProcessorRegisteredToService()) {
+      deregister();
+    }
+  }
+
+  private void deregister() {
+    slackConnectionService.deregisterProcessor(this);
+  }
+
+  @Override
+  public void onTrigger(ProcessContext context, ProcessSessionFactory sessionFactory) throws ProcessException {
+    if (processSessionFactory == null) {
+      processSessionFactory = sessionFactory;
+    }
+    if (!isProcessorRegisteredToService()) {
+      registerProcessorToService(context);
+    }
+    context.yield();
+  }
+
+  private boolean isProcessorRegisteredToService() {
+    return slackConnectionService != null
+      && slackConnectionService.isProcessorRegistered(this);
+  }
+
+  private void registerProcessorToService(ProcessContext context) {
+      try {
+        slackConnectionService = context.getProperty(SLACK_CONNECTION_SERVICE)
+          .asControllerService(SlackConnectionService.class);
+        slackConnectionService.registerProcessor(this, getMessageHandler(processSessionFactory));
+      } catch (Exception e) {
+        getLogger().error("Error while creating slack client", e);
+        slackConnectionService.deregisterProcessor(this);
+        context.yield();
+      }
+  }
+
+  private Consumer<String> getMessageHandler(ProcessSessionFactory sessionFactory) {
+    return message -> {
+      ProcessSession session = sessionFactory.createSession();
+      try {
+        FlowFile flowFile = session.create();
+        session.write(flowFile, outputStream -> outputStream.write(message.getBytes()));
+        session.getProvenanceReporter().receive(flowFile, "slack");
+
+        if (matches(message)) {
+          session.transfer(flowFile, MATCHED_MESSAGES_RELATIONSHIP);
+        } else {
+          session.transfer(flowFile, UNMATCHED_MESSAGES_RELATIONSHIP);
+        }
+        session.commit();
+      } catch (Exception e) {
+        session.rollback();
+      }
+    };
+  }
+
+  private boolean matches(String message) {
+    if (matchAny) {
+      return true;
+    }
+
+    try (JsonReader reader = Json.createReader(new StringReader(message))) {
+      return Optional.ofNullable(reader.readObject())
+        .map(jsonObject -> jsonObject.getString(JSON_OBJECT_TYPE_KEY, null))
```

Comment:
Can we put this TYPE_KEY value to a FlowFile attribute, too? I think it will be helpful for routing messages at downstream flow.